File: lab/Taller 6/diffusionEquation.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>


float func_initial(float x);
void copy(double *new, double *old, int length);

int main(){
  int points = 50;
  double n_time = 1500;
  double c = 1.0;
  double sigma = 0.2;
  double nu = 0.3;
  double dx=1.0/points;
  double dt = sigma*dx*dx/nu;
  double alpha = dx*nu ;
  double *u_past = malloc(points*sizeof(double));
  u_past[0]=0.0;
  u_past[points-1]=0.0;
  int i,j;
  FILE *out = fopen("data_c.dat","w");

  for(i=1; i<points-1;i++){
    u_past[i]=func_initial(i*dx);
  }

  fprintf(out, "%f ", 0.0);
  for(i=0; i< points; i++){
    fprintf(out, "%f ", u_past[i]);
  }
  fprintf(out, "\n");

  double *u = malloc(points*sizeof(double));

  for (j=2; j<n_time; j++){
    for(i=0; i<points-1; i++){
      u[i] = nu * alpha * u_past[i+1]  + (1.0 - 2.0*nu*alpha)*u_past[i] + nu*alpha*u_past[i-1];
    }
    copy (u_past, u ,points);
    fprintf(out, "%f ", dt*j);
    for(i=0; i< points; i++){
      fprintf(out, "%f ", u[i]);
    }
    fprintf(out, "\n");
  }

  return 0;
}

float func_initial(float x){
  if (x>0.33 && x<0.66)
    return 1.0;
  else 
    return 0.0;
}

void copy(double *new, double *old, int length){
  int i;
  for (i=0; i<length; i++)
    new[i]=old[i];
}

File: lab/Taller 6/diffusionEquation.py
import numpy as np


n_x = 80
n_t = 700

c = 1.0
nu = 0.3
sigma = 0.2

x = np.linspace(0, 1.0, n_x)
dx = x[1]-x[0]


dt = sigma*dx**2/nu
alpha = dt/dx**2

u = np.zeros(n_x)

u[np.where((x<0.66) & (x>0.33))] = 1.0

fileout= open ("data_python.dat", "w")

for n in range(n_t):
    u_past = u.copy() 
    fileout.write("%f " % (n*dt))
    for i in range(1,n_x-1):
        u[i] = nu * alpha * u_past[i+1]  + (1.0 - 2.0*nu*alpha)*u_past[i] + nu*alpha*u_past[i-1]
        fileout.write("%f " % (u[i]))
    fileout.write("\n")
                      
        
